File: RemoteBees-iOS/RemoteBeesService/Managers/RemoteBeesServiceConfiguration.swift
//
//  RemoteBeesServiceConfiguration.swift
//  RemoteBees
//

import Foundation
import HTTPServiceKit

public enum RemoteBeesServiceEnvironment: String {
    case development
    case staging
    case production
    
    var remotiveUrl: URL {
        switch self {
            case .development, .staging, .production:
                return URL(string: "https://remotive.io/")!
        }
    }
}

public struct RemoteBeesServiceConfiguration {
    public let environment: RemoteBeesServiceEnvironment
    public let localeIdentifier: String
    public let requestTimeoutInterval: TimeInterval
    public let logger: HTTPServiceLogger?
    
    public init(
        environment: RemoteBeesServiceEnvironment,
        localeIdentifier: String,
        requestTimeoutInterval: TimeInterval = 30.0,
        logger: HTTPServiceLogger? = HTTPServiceConsoleLogger()
    ) {
        self.environment = environment
        self.localeIdentifier = localeIdentifier
        self.requestTimeoutInterval = requestTimeoutInterval
        self.logger = logger
    }
}

extension RemoteBeesServiceConfiguration {
    private static let decoders: [String: HTTPServiceDecoder] = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"  // Remotive date format
        let jsonDecoder = JSONDecoder()
        jsonDecoder.dateDecodingStrategy = .formatted(formatter)

        return [
            "application/xml": jsonDecoder,
            "application/json": jsonDecoder,
        ]
    }()

    func content() -> HTTPService.Config {
        var httpServiceConfig = HTTPService.Config(baseUrl: nil)
        httpServiceConfig.requestTimeoutInterval = self.requestTimeoutInterval
        httpServiceConfig.headers.merge(["Cache-Control": "no-cache"]) { (_, new) in new }
        httpServiceConfig.cacheStore = URLCache.shared
        return httpServiceConfig
    }

    func remotive() -> HTTPService.Config {
        var httpServiceConfig = HTTPService.Config(baseUrl: self.environment.remotiveUrl)
        
        // Configure headers as needed. For example
        var headers = [String: String]()

        /*
        if !self.environment.appToken.isEmpty {
            headers["AppToken"] = self.environment.appToken
        }
        if let session = session() {
            headers["UserToken"] = session.authUserToken
        }
        */

        headers["Accept-Language"] = self.localeIdentifier
        headers["Cache-Control"] = "no-cache"

        httpServiceConfig.headers.merge(headers) { (_, new) in new }
        httpServiceConfig.decoders = Self.decoders
        httpServiceConfig.logger = self.logger
        httpServiceConfig.cacheStore = URLCache.shared
        httpServiceConfig.requestTimeoutInterval = self.requestTimeoutInterval

        return httpServiceConfig
    }
    
}

File: RemoteBees-Android/RemoteBeesService/src/main/java/dev/beehive/remotebees/service/manager/RemoteBeesServiceConfiguration.kt
/***************************************************************************
 * This source file is part of the RemoteBees open source project.         *
 *                                                                         *
 * Licensed under the MIT License. See LICENSE for license information     *
 ***************************************************************************/

package dev.beehive.remotebees.service.manager

import android.content.Context
import com.inmotionsoftware.httpservice.HTTPService
import com.inmotionsoftware.httpservice.HTTPServiceConsoleLogger
import com.inmotionsoftware.httpservice.HTTPServiceLogger
import com.inmotionsoftware.httpservice.TimeInterval
import com.inmotionsoftware.httpservice.cache.CacheStore
import com.inmotionsoftware.httpservice.cache.MemDiskLruCacheStore
import dev.beehive.remotebees.service.BuildConfig
import java.lang.ref.WeakReference
import java.net.URL

enum class RemoteBeesServiceEnvironment {
    Development,
    Staging,
    Production
    ;

    val remotiveUrl: URL
        get() = when (this) {
            Development, Staging, Production -> URL("https://remotive.io/")
        }
}

data class RemoteBeesServiceConfiguration(
    val context: WeakReference<Context>,
    val environment: RemoteBeesServiceEnvironment,
    val localeIdentifier: String,
    val requestTimeoutInterval: TimeInterval = 15,
    val enableHttpLogging: Boolean = BuildConfig.DEBUG,
    val logger: HTTPServiceLogger? = HTTPServiceConsoleLogger
)

// Extension

internal fun RemoteBeesServiceConfiguration.content(): HTTPService.Config {
    val httpServiceConfig = this.defaultConfig()

    val headers = HashMap<String, String>()
    headers["Cache-Control"] = "no-cache"
    httpServiceConfig.headers = headers

    return httpServiceConfig
}

internal fun RemoteBeesServiceConfiguration.remotive(): HTTPService.Config {
    val httpServiceConfig = this.defaultConfig(baseUrl = URL("${this.environment.remotiveUrl}"))

    val headers = HashMap<String, String>()
    headers["Accept-Language"] = this.localeIdentifier
    headers["Cache-Control"] = "no-cache"
    httpServiceConfig.headers = headers

    return httpServiceConfig
}

private fun RemoteBeesServiceConfiguration.cacheStore(context: Context): CacheStore {
    val cacheDir = context.cacheDir
    val diskCacheSize = 50 * 1024 * 1024
    val memCacheSize = 2 * 1024 * 1024
    return MemDiskLruCacheStore(cacheDir, diskCacheSize.toLong(), memCacheSize)
}

private fun RemoteBeesServiceConfiguration.defaultConfig(baseUrl: URL? = null): HTTPService.Config {
    val httpServiceConfig = HTTPService.Config(baseUrl)

    httpServiceConfig.cacheStore = this.context.get()?.let(this::cacheStore)
    httpServiceConfig.isAlwaysTrustHost = BuildConfig.DEBUG
    httpServiceConfig.requestTimeoutInterval = this.requestTimeoutInterval
    httpServiceConfig.logger = this.logger
    return httpServiceConfig
}
